pkg/agent/proxy/proxier.go

```diff
@@ -136,65 +136,66 @@ func (p *proxier) isInitialized() bool {
-				if err := p.uninstallNodePortService(uint16(svcInfo.NodePort()), svcInfo.OFProtocol); err != nil {
-					klog.ErrorS(err, "Failed to remove flows and configurations of Service", "Service", svcPortName)
-					continue
-				}
+		// Remove associated Endpoints flows.
+		if endpoints, ok := p.endpointsInstalledMap[svcPortName]; ok {
+			if err := p.removeStaleEndpoints(endpoints, svcInfo.Protocol()); err != nil {
+				klog.ErrorS(err, "Error when removing Endpoints flows for Service", "ServiceName", svcPortName, "ServiceInfo", svcInfoStr)
+				continue
+			}
+			delete(p.endpointsInstalledMap, svcPortName)
+		}
+		// Remove NodePort flows and configurations.
+		if p.proxyAll && svcInfo.NodePort() > 0 {
+			if err := p.uninstallNodePortService(uint16(svcInfo.NodePort()), svcInfo.OFProtocol); err != nil {
+				klog.ErrorS(err, "Error when uninstalling NodePort flows and configurations for Service", "ServiceName", svcPortName, "ServiceInfo", svcInfoStr)
+				continue
 			}
 		}
 		// Remove LoadBalancer flows and configurations.
 		if p.proxyLoadBalancerIPs && len(svcInfo.LoadBalancerIPStrings()) > 0 {
 			if err := p.uninstallLoadBalancerService(svcInfo.LoadBalancerIPStrings(), uint16(svcInfo.Port()), svcInfo.OFProtocol); err != nil {
-				klog.ErrorS(err, "Failed to remove flows and configurations of Service", "Service", svcPortName)
+				klog.ErrorS(err, "Error when uninstalling LoadBalancer flows and configurations for Service", "ServiceName", svcPortName, "ServiceInfo", svcInfoStr)
 				continue
 			}
 		}
-		// Remove Service group whose Endpoints are local.
-		if svcInfo.ExternalPolicyLocal() {
-			if groupIDLocal, exist := p.groupCounter.Get(svcPortName, true); exist {
-				if err := p.ofClient.UninstallServiceGroup(groupIDLocal); err != nil {
-					klog.ErrorS(err, "Failed to remove Group of local Endpoints for Service", "Service", svcPortName)
-					continue
-				}
-				p.groupCounter.Recycle(svcPortName, true)
+		// Remove Service group which has only local Endpoints.
+		if groupID, exist := p.groupCounter.Get(svcPortName, true); exist {
+			if err := p.ofClient.UninstallServiceGroup(groupID); err != nil {
+				klog.ErrorS(err, "Error when uninstalling group of local Endpoints for Service", "ServiceName", svcPortName, "ServiceInfo", svcInfoStr)
+				continue
 			}
+			p.groupCounter.Recycle(svcPortName, true)
 		}
 		// Remove Service group which has all Endpoints.
 		if groupID, exist := p.groupCounter.Get(svcPortName, false); exist {
 			if err := p.ofClient.UninstallServiceGroup(groupID); err != nil {
-				klog.ErrorS(err, "Failed to remove Group of all Endpoints for Service", "Service", svcPortName)
+				klog.ErrorS(err, "Error when uninstalling group of all Endpoints for Service", "ServiceName", svcPortName, "ServiceInfo", svcInfoStr)
 				continue
 			}
 			p.groupCounter.Recycle(svcPortName, false)
 		}
 
 		delete(p.serviceInstalledMap, svcPortName)
-		p.deleteServiceByIP(svcInfo.String())
+		p.deleteServiceByIP(svcInfoStr)
```

Comment:
How about `deleteServiceByName`?

Reply:
How about `deleteService` since we do not have other delete functions by a filter? Name is not accurate as the parameter is `ClusterIP:Port/protocol`

Reply:
I'll change this in another PR.